File: final/src/symbols/Environment.java
package symbols;

import errors.SemanticError;
import java.util.Hashtable;

public class Environment {
	private Hashtable<String, Object> table;
	private Environment previous;
	
	public Environment() {
		table = new Hashtable<String, Object>();
		previous = null;
	}
	
	public Environment(Environment prev) {
		table = new Hashtable<String, Object>();
		previous = prev;
	}
	
	public Environment getPrevious() {
		return previous;
	}
	
	public void setPrevious(Environment previous) {
		this.previous = previous;
	}
	
	public void insert(String varName) throws SemanticError{
		if (table.get(varName) == null) {
			table.put(varName, Double.valueOf(0.0));
		} else {
			throw new SemanticError("The variable was already defined");
		}
	}
	
	public Object lookup(String varName) {
		Object found;
		for (Environment env = this; env.previous != null; env = env.previous) {
			found = env.table.get(varName);
			if (found != null) {
				return found;
			}
		}
		return null;
	}
	
	public void set(String varName, Object value) throws SemanticError{
		Object found = null;
		for (Environment env = this; env.previous != null; env = env.previous) {
			found = env.table.get(varName);
			if (found != null) {
				env.table.put(varName, value);
				break;
			}
		}
		if (found == null) {
			throw new SemanticError("The variable has not been defined");
		}
	}
}

File: stage-3/Environment.py
class Environment:
    __table = {}
    __previous = None

    def __init__(self, env = None):
        self.__previous = env
        self.__table = {}

    def getPrevious(self):
        return self.__previous
    
    def lookup(self, variable):
        if variable in self.__table:
            return self.__table[variable]
        elif self.__previous != None:
            return self.__previous.lookup(variable)
        else: 
            return None
    
    def insert(self, variable):
        if not variable in self.__table:
            self.__table[variable] = 0.0
            return True
        else:
            return False
        
    def set(self, variable, value):
        env = self
        found = False
        while env != None:
            if variable in env.__table:
                found = True
                break
            env = env.__previous
        
        if not found:
            env.__table[variable] = value
            return True
        else: 
            # return False
            raise Exception('Semantic Error - The variable has not been defined')
